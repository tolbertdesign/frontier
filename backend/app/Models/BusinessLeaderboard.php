<?php

namespace App\Models;

use App\Entities\SponsorType;
use App\Entities\Pledge;
use App\Entities\PledgeType;
use App\Entities\Program;
use DB;

class BusinessLeaderboard
{
    private $program;

    public function __construct(Program $program)
    {
        $this->program = $program;
    }

    public function getPledges()
    {
        $businessPledgeTypeId = SponsorType::where('sponsor_type', 'Business')->first()->id;
        $businessPledges   = Pledge::where(
            [
                'program_id'       => $this->program->id,
                'sponsor_type_id'  => $businessPledgeTypeId,
                'anon'             => '0',
                'pledges.deleted'  => '0'
            ]
        )

            ->join('users', 'users.id', '=', 'pledges.participant_user_id')
            ->join('pledge_types', 'pledge_types.id', '=', 'pledges.pledge_type_id')
            ->join('programs', 'programs.id', '=', 'pledges.program_id')
            ->select(
                [
                    DB::raw('min(pledges.program_id) AS program_id'),
                    DB::raw('min(pledges.business_name) AS business_name'),
                    DB::raw('min(pledges.business_website) AS business_website'),
                    DB::raw('min(pledges.comment) AS comment'),
                    DB::raw('min(pledges.pledge_type_id) AS pledge_type_id'),
                    DB::raw('min(pledges.show_comment) AS show_comment'),
                    DB::raw('min(users.laps) AS laps'),
                    DB::raw('sum(amount) AS amount'),
                    DB::raw('
                        SUM(IF(`pledge_types`.`id` = ' . PledgeType::FLAT . ',
                            `pledges`.`amount`,
                            `pledges`.`amount` * IF(users.laps = 0 OR users.laps IS NULL,
                            `programs`.`unit_flat_conversion`,
                            `users`.`laps`))) AS total_est
                        '),
                ]
            )
            ->groupBy(
                [
                    DB::raw('COALESCE(`pledges`.`family_pledge_id`, `pledges`.`id`)')
                ]
            )
            ->orderBy('total_est', 'desc')
            ->get();

        return $businessPledges;
    }
}
