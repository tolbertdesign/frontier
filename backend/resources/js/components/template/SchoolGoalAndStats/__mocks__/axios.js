export default {
  get: jest.fn(() => Promise.resolve({ data: 5 })),
  all: jest.fn(() => Promise.resolve({ data: true })),
}
