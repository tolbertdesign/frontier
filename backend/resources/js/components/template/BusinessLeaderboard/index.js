import BusinessLeaderboard from './BusinessLeaderboard'

export default BusinessLeaderboard
