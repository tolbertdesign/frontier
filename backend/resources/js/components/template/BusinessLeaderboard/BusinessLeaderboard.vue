<template>
  <div class="business-leaderboard card-content py-4 px-7">
    <p class="mb-8">
      {{ business_leaderboard_heading_1 }} <a
        :href="`/v3/tkdashboard/?redirect=auth/login/${fr_code}/view-participant/pledge`"
        class="font-bold text-secondary"
      >{{ lang.business_leaderboard_heading_2 }}</a> {{ lang.business_leaderboard_heading_3 }}
    </p>

    <table class="table is-fullwidth -mx-3 mb-8">
      <tbody v-if="hasBusinessPledges">
        <tr
          v-for="(businessPledge, index) in businessPledgesToShow"
          :key="businessPledge.id"
        >
          <td>
            <h3>
              {{ index + 1 }}.
              <a
                v-if="businessPledge.business_website"
                :href="businessPledge.business_website"
                target="_blank"
              >{{ businessPledge.business_name }}</a>
              <span v-else>{{ businessPledge.business_name }}</span>
            </h3>
            <h4 v-if="businessPledge.comment && businessPledge.show_comment == 1">
              {{ businessPledge.comment }}
            </h4>
          </td>
          <td class="text-right whitespace-no-wrap">
            {{ pledgeTotal(businessPledge) }}
          </td>
        </tr>
      </tbody>
      <tbody v-else>
        <tr>
          <td>
            <h3>{{ lang.top_business_pledge }}</h3>
            <h4>{{ lang.note_goes_here }}</h4>
          </td>
          <td class="text-right">
            {{ lang.pledge_placeholder_amount }}
          </td>
        </tr>
      </tbody>
    </table>

    <div
      v-if="shouldShowViewMore"
      class="text-center"
    >
      <button
        class="button is-white has-text-secondary has-text-weight-bold"
        @click.prevent="toggle"
      >
        {{ showAll ? lang.show + ' ' + lang.less : lang.show + ' ' + lang.more }}
      </button>
    </div>
  </div>
</template>

<script>
import currency from '@/filters/currency'
import axios from 'axios'
const VIEW_LESS_NUMBER = 5

export default {
  name: 'BusinessLeaderboard',
  status: 'prototype',
  release: '1.0.0',
  props: {
    program: {
      type: Object,
      default: null,
    },
  },
  data () {
    return {
      showAll: false,
      businessPledges: [],
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    fr_code () {
      return this.program.participants.find(participant => participant.deleted !== 1).fr_code
    },
    business_leaderboard_heading_1 () {
      return this.parseLanguage(
        this.lang.business_leaderboard_heading_1,
        {
          event_name: this.program.event_name,
        },
      )
    },
    hasBusinessPledges () {
      return this.businessPledges.length > 0
    },
    businessPledgesToShow () {
      return this.showAll ? this.businessPledges : this.businessPledges.slice(0, VIEW_LESS_NUMBER)
    },
    shouldShowViewMore () {
      return this.businessPledges.length > VIEW_LESS_NUMBER
    },
  },
  created () {
    axios.get(`/v3/api/business-leaderboard-pledges/${this.program.id}`)
      .then(response => {
        this.businessPledges = response.data
      }).catch(error => {
        console.error(error)
      })
  },
  methods: {
    toggle () {
      this.showAll = !this.showAll
    },
    pledgeTotal (pledge) {
      const totals = { low: 0, high: 0 }
      if (pledge.laps) {
        totals.low = pledge.total_est
        totals.high = pledge.total_est
      } else if (pledge.pledge_type_id === 1) {
        totals.low = parseFloat(pledge.amount) * this.program.unit_range_low
        totals.high = parseFloat(pledge.amount) * this.program.unit_range_high
      } else {
        totals.low = parseFloat(pledge.amount)
        totals.high = parseFloat(pledge.amount)
      }
      let totalsString = currency(totals.low, true, '$')
      if (totals.low !== totals.high) {
        totalsString += ' to ' + currency(totals.high, true, '$')
      }
      return totalsString
    },
  },
}
</script>
