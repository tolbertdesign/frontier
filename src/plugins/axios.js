import Vue from 'vue'
import axios from 'axios'
import Swal from 'sweetalert2'
import store from '@/vuex/store'
import router from '@/router'
import i18n from '@/plugins/i18n'

const baseURL = 'http://tolbert.test'

const instance = axios.create({
  baseURL: baseURL,
  headers: {
    'X-Requested-With': 'XMLHttpRequest',
  },
})
instance.defaults.withCredentials = true

// Request interceptor
instance.interceptors.request.use(request => {
  const token = store.getters['auth/token']
  if (token) {
    request.headers.common.Authorization = `Bearer ${token}`
  }

  const locale = store.getters['lang/locale']
  if (locale) {
    request.headers.common['Accept-Language'] = locale
  }

  // request.headers['X-Socket-Id'] = Echo.socketId()

  return request
})

// Response interceptor
instance.interceptors.response.use(response => response, error => {
  const { status } = error.response

  if (status >= 500) {
    Swal.fire({
      type: 'error',
      title: i18n.t('error_alert_title'),
      text: i18n.t('error_alert_text'),
      reverseButtons: true,
      confirmButtonText: i18n.t('ok'),
      cancelButtonText: i18n.t('cancel'),
    })
  }

  if (status === 401 && store.getters['auth/check']) {
    Swal.fire({
      type: 'warning',
      title: i18n.t('token_expired_alert_title'),
      text: i18n.t('token_expired_alert_text'),
      reverseButtons: true,
      confirmButtonText: i18n.t('ok'),
      cancelButtonText: i18n.t('cancel'),
    }).then(() => {
      store.commit('auth/LOGOUT')

      router.push({ name: 'login' })
    })
  }

  return Promise.reject(error)
})

Vue.use({
  install (Vue) {
    Vue.prototype.$http = instance
  },
})

export default instance
