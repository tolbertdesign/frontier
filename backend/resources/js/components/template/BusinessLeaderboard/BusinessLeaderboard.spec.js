import { shallowMount, createLocalVue } from '@vue/test-utils'
import BusinessLeaderboard from './BusinessLeaderboard.vue'
import Vuex from 'vuex'
import axios from 'axios'
import LanguageParser from '@/mixins/LanguageParser'
jest.mock('axios')

function fakePledge (num) {
  return {
    id: num,
    program_id: 1,
    business_website: 'www.test' + num + '.com',
    business_name: 'Test' + num,
    comment: num % 2 === 0 ? 'Comment' : '',
    pledge_type_id: num % 2 === 0 ? 1 : 3,
    show_comment: num % 2 === 0 ? 1 : 0,
    laps: num % 2 === 0 ? num : 0,
    amount: 20,
    total_est: 100,
  }
}

const businessPledges = []
for (let i = 1; i <= 6; i++) {
  businessPledges.push(fakePledge(i))
}

axios.get.mockImplementation((url) => {
  if (url.startsWith('/v3/api/business-leaderboard-pledges')) {
    return Promise.resolve({ data: businessPledges })
  }
})

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.filter('currency', jest.fn())
localVue.mixin(LanguageParser)

describe('BusinessLeaderboard', () => {
  let program = {}
  let store
  let storeProps

  beforeEach(() => {
    program = {
      id: 1,
      name: 'Program Name',
      unit_range_low: 1,
      unit_range_high: 2,
      participants: [
        {
          fr_code: 'FFFFFFFF',
        },
      ],
    }
    storeProps = {
      lang: {
        business_leaderboard_heading_1: 'heading 1',
        business_leaderboard_heading_2: 'heading 2',
        business_leaderboard_heading_3: 'heading 3',
        view: 'View',
        show: 'Show',
        less: 'Less',
        more: 'More',
        top_business_pledge: 'top pledge',
        note_goes_here: 'note goes here',
      },
    }
    store = new Vuex.Store({
      state: storeProps,
    })
  })

  it('is a Vue instance', () => {
    const wrapper = shallowMount(BusinessLeaderboard, {
      localVue,
      store,
      propsData: { program },
    })
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('renders correctly', done => {
    const wrapper = shallowMount(BusinessLeaderboard, {
      localVue,
      store,
      propsData: { program },
    })
    wrapper.vm.$nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot()
      done()
    })
  })
})
