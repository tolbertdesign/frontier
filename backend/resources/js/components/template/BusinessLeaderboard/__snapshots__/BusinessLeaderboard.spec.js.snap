// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessLeaderboard renders correctly 1`] = `
<div class="business-leaderboard card-content py-4 px-7">
  <p class="mb-8">
    heading 1 <a href="/v3/tkdashboard/?redirect=auth/login/FFFFFFFF/view-participant/pledge" class="font-bold text-secondary">heading 2</a> heading 3
  </p>
  <table class="table is-fullwidth -mx-3 mb-8">
    <tbody>
      <tr>
        <td>
          <h3>
            1.
            <a href="www.test1.com" target="_blank">Test1</a></h3>
          <!---->
        </td>
        <td class="text-right whitespace-no-wrap">
          $20
        </td>
      </tr>
      <tr>
        <td>
          <h3>
            2.
            <a href="www.test2.com" target="_blank">Test2</a></h3>
          <h4>
            Comment
          </h4>
        </td>
        <td class="text-right whitespace-no-wrap">
          $100
        </td>
      </tr>
      <tr>
        <td>
          <h3>
            3.
            <a href="www.test3.com" target="_blank">Test3</a></h3>
          <!---->
        </td>
        <td class="text-right whitespace-no-wrap">
          $20
        </td>
      </tr>
      <tr>
        <td>
          <h3>
            4.
            <a href="www.test4.com" target="_blank">Test4</a></h3>
          <h4>
            Comment
          </h4>
        </td>
        <td class="text-right whitespace-no-wrap">
          $100
        </td>
      </tr>
      <tr>
        <td>
          <h3>
            5.
            <a href="www.test5.com" target="_blank">Test5</a></h3>
          <!---->
        </td>
        <td class="text-right whitespace-no-wrap">
          $20
        </td>
      </tr>
    </tbody>
  </table>
  <div class="text-center"><button class="button is-white has-text-secondary has-text-weight-bold">
      Show More
    </button></div>
</div>
`;
